rules/rbe_repo/rbe_repo_spec.bzl

```diff
@@ -0,0 +1,435 @@
+    "rbe_repo_gen_spec",
+)
+
+_SEPARATOR = ":::"
+
+REPO_SPEC_KEYS = [
+    "container_repo",
+    "container_registry",
+    "output_base",
+    "repo_name",
+    "rbe_repo_gen_spec",
+]
+
+CONFIG_SPEC_FIELDS = [
+    "bazel_to_config_spec_names_map",
+    "container_to_config_spec_names_map",
+    "default_toolchain_config_spec",
+    "latest_container",
+    "toolchain_config_specs",
+]
+
+def default_rbe_repo_spec():
+    return {
+        "repo_name": "bazel_toolchains",
+        "output_base": "configs/ubuntu16_04_clang",
+        "container_repo": "google/rbe-ubuntu16-04",
+        "container_registry": "marketplace.gcr.io",
+        "rbe_repo_gen_spec": rbe_repo_gen_spec(),
+    }
+
+def validate_rbe_repo_spec(name, rbe_repo_spec):
+    """ Validates the given rbe_repo_spec is 
+
+Should only be called from the rbe_autoconfig macro
+in //rules/rbe_repo.bzl
+
+    Args:
+      name: Name of the rbe_autoconfig rule
+      rbe_repo_spec: The rbe_repo_spec to validate
+
+    """
+    _validate_repo_spec_keys(name, rbe_repo_spec)
+    _validate_rbe_repo_gen_spec(name, rbe_repo_spec)
+
+# Validates the rbe_repo_spec is a struct and has all required top level keys
+def _validate_repo_spec_keys(name, rbe_repo_spec):
+    if type(rbe_repo_spec) != "dict":
+        fail("rbe_repo_spec in %s is not a dict: '%s'" % (name, rbe_repo_spec))
+    for key in REPO_SPEC_KEYS:
+        if not rbe_repo_spec.get(key):
+            fail("rbe_repo_spec in %s does not contain required key '%s'" % (name, key))
+    for key in rbe_repo_spec.keys():
+        if key not in REPO_SPEC_KEYS:
+            fail("rbe_repo_spec in %s contain unnecessary key '%s'" % (name, key))
+
+def _validate_rbe_repo_gen_spec(name, rbe_repo_spec):
+    if str(type(rbe_repo_spec["rbe_repo_gen_spec"])) != "struct":
+        fail(("%s has a rbe_repo_spec[\"rbe_repo_gen_spec\"] that is not a struct " +
+              "'%s' was passed as value.") %
+             (name, rbe_repo_spec["rbe_repo_gen_spec"]))
+
+    # validate the rbe_repo_gen_spec object is a struct with all required fields
+    for field in CONFIG_SPEC_FIELDS:
+        if not hasattr(rbe_repo_spec["rbe_repo_gen_spec"], field):
+            fail(("rbe_repo_spec[\"rbe_repo_gen_spec\"] in %s does not contain " +
+                  "required field '%s'") % (name, field))
+    _validate_default_toolchain_config_spec(name, rbe_repo_spec)
+    _validate_bazel_to_config_spec_names_map(name, rbe_repo_spec)
+    _validate_container_to_config_spec_names_map(name, rbe_repo_spec)
+    _validate_configs(name, rbe_repo_spec)
+
+def _validate_default_toolchain_config_spec(name, rbe_repo_spec):
+    if (rbe_repo_spec["rbe_repo_gen_spec"].default_toolchain_config_spec != "" and
+        type(rbe_repo_spec["rbe_repo_gen_spec"].default_toolchain_config_spec) != "struct"):
+        fail(("%s has a rbe_repo_spec[\"rbe_repo_gen_spec\"] field 'default_toolchain_config_spec' " +
+              "that is not either an empty string or a 'struct'. " +
```

Comment:
Suggest using neither nor instead of "not either"

Reply:
fixed